packages/web-client/app/models/workflow/workflow-session.ts

```diff
@@ -60,12 +114,98 @@ export default class WorkflowSession {
     );
   }
 
+  getValue<T extends SupportedType>(key: string): T | null {
+    const data: string | null = this._state[key];
+    if (data !== null && data !== undefined) {
+      let json = JSON.parse(data);
+      if (json.type === 'Date') {
+        return new Date(json.value) as T;
+      } else if (json.type === 'BN') {
+        return new BN(json.value) as T;
+      } else {
+        return json.value;
+      }
+    }
+    return null;
+  }
+
+  getValues(): WorkflowSessionDictionary {
+    let result: WorkflowSessionDictionary = {};
+    for (const key in this._state) {
+      const data: string | null = this._state[key];
+      if (data !== null && data !== undefined) {
+        let json = JSON.parse(data);
+        if (json.type === 'Date') {
+          result[key] = new Date(json.value);
+        } else if (json.type === 'BN') {
+          result[key] = new BN(json.value);
+        } else {
+          result[key] = json.value;
+        }
+      }
+    }
+    return result;
+  }
+
+  setValue(key: string, value: SupportedType): void;
+  setValue(hash: Record<string, SupportedType>): void;
+  setValue(
+    hashOrKey: Record<string, SupportedType> | string,
+    value?: SupportedType
+  ): void {
+    if (typeof hashOrKey === 'string') {
+      this.setStateProperty(hashOrKey, value!);
+    } else {
+      for (const key in hashOrKey) {
+        this.setStateProperty(key, hashOrKey[key]);
+      }
+    }
+    // eslint-disable-next-line no-self-assign
+    this._state = this._state; // for reactivity
+    this.persistToStorage();
+  }
+
+  private setStateProperty(key: string, value: SupportedType) {
+    serializeToState(this._state, key, value);
+  }
+
   private persistToStorage(): void {
     if (!this.hasPersistence) return;
 
     this.workflow?.workflowPersistence.persistData(
       this.workflow.workflowPersistenceId,
-      { name: this.workflow.name, state: this.state }
+      { name: this.workflow.name, state: this._state }
     );
   }
 }
+
+export function serializeToState(
+  state: Record<string, string>,
+  key: string,
+  value: SupportedType
+) {
+  if (value instanceof Date) {
+    state[key] = JSON.stringify({
+      value: value.toISOString(),
+      type: 'Date',
+    });
+  } else if (value instanceof BN) {
```

Comment:
Ah, interesting. This seems like a good suggestion. Is there a way to create a failing test for this I wonder?

Reply:
Seems somewhat dependent on the implementation of BN.isBN (https://github.com/indutny/bn.js/blob/5d5532d9fb7192b5d2545bdb2aa024ca1822a313/lib/bn.js):

```
BN.isBN = function isBN (num) {
    if (num instanceof BN) {
      return true;
    }

    return num !== null && typeof num === 'object' &&
      num.constructor.wordSize === BN.wordSize && Array.isArray(num.words);
  };
  ```

Reply:
We could create `num` where ` num !== null && typeof num === 'object' && num.constructor.wordSize === BN.wordSize && Array.isArray(num.words)`, and ensure that this is serialized and deserialized as a BN. 

It is curious why `instanceof` would fail though, unless we are somehow using multiple different BN versions?

Reply:
Perhaps related to this: https://raganwald.com/2014/01/14/the-new-javascript-problem.html